File: web/lib/doc/render.ts
/**
 * Client-side AST → HTML renderer.  Mirror of api/internal/generate/doc/render.go.
 *
 * This runs in the browser for live-preview updates (no server round-trip
 * for every keystroke).  It MUST produce output byte-compatible with the
 * server renderer — otherwise authors see flicker between the in-editor
 * preview and the post-save rendered HTML.  The test suite in render.test.ts
 * pins this invariant with paired fixtures.
 */

import type {
  Block,
  Condition,
  Doc,
  FieldFormat,
  Inline,
  Mark,
  TableRow,
} from "./ast";
import {
  evalCondition,
  formatField,
  iterItems,
  pushScope,
  resolvePath,
  rootScope,
  type Scope,
} from "./eval";

export interface Diagnostic {
  nodeId?: string;
  severity: "error" | "warning";
  message: string;
}

export interface RenderOptions {
  /** Data map used to resolve field references. */
  data?: Record<string, unknown>;
  /** Doc-level locale for formatters (currency, dates). */
  locale?: string;
}

export interface RenderResult {
  html: string;
  diagnostics: Diagnostic[];
}

/** Render `doc` to inner-body HTML.  The caller decides whether to wrap in
 *  a document shell (usually the preview iframe does via wrapInDocument). */
export function renderDoc(doc: Doc, opts: RenderOptions = {}): RenderResult {
  const ctx: RenderContext = {
    locale: opts.locale ?? "en-US",
    diagnostics: [],
  };
  const scope = rootScope(opts.data ?? {});
  const parts: string[] = [];
  for (const b of doc.content) {
    parts.push(renderBlock(b, scope, ctx));
  }
  return { html: parts.join(""), diagnostics: ctx.diagnostics };
}

interface RenderContext {
  locale: string;
  diagnostics: Diagnostic[];
}

function pushDiag(
  ctx: RenderContext,
  severity: "error" | "warning",
  message: string,
  nodeId?: string,
): void {
  ctx.diagnostics.push({ severity, message, nodeId });
}

// ---------------------------------------------------------------------------
// Block rendering
// ---------------------------------------------------------------------------

function renderBlock(b: Block, scope: Scope, ctx: RenderContext): string {
  switch (b.type) {
    case "heading":
      return renderHeading(b, scope, ctx);
    case "paragraph":
      return renderParagraph(b, scope, ctx);
    case "bulletList":
      return renderList(b.items, "ul", b.id, scope, ctx);
    case "orderedList":
      return renderList(b.items, "ol", b.id, scope, ctx);
    case "table":
      return renderTable(b.rows, b.id, scope, ctx);
    case "divider":
      return "<hr />\n";
    case "image":
      return renderImage(b, ctx);
    case "repeat":
      return renderRepeat(b, scope, ctx);
    case "if":
      return renderIf(b, scope, ctx);
    case "raw":
      return b.html;
    default: {
      // Exhaustive-check escape: if a new block type is added, TS will
      // flag this branch at compile time.
      const _never: never = b;
      void _never;
      return "";
    }
  }
}

function renderHeading(
  b: { level: number; content: Inline[]; id?: string },
  scope: Scope,
  ctx: RenderContext,
): string {
  const lvl = Math.min(6, Math.max(1, b.level || 1));
  return `<h${lvl}${nodeAttr(b.id)}>${renderInlines(b.content, scope, ctx)}</h${lvl}>\n`;
}

function renderParagraph(
  b: { content: Inline[]; align?: string; id?: string },
  scope: Scope,
  ctx: RenderContext,
): string {
  let style = "";
  if (b.align === "center") style = ' style="text-align:center"';
  else if (b.align === "right") style = ' style="text-align:right"';
  else if (b.align === "left") style = ' style="text-align:left"';
  return `<p${nodeAttr(b.id)}${style}>${renderInlines(b.content, scope, ctx)}</p>\n`;
}

function renderList(
  items: { id?: string; content: Block[] }[],
  tag: "ul" | "ol",
  id: string | undefined,
  scope: Scope,
  ctx: RenderContext,
): string {
  const parts: string[] = [`<${tag}${nodeAttr(id)}>\n`];
  for (const it of items) {
    parts.push(`<li${nodeAttr(it.id)}>`);
    if (it.content.length === 1 && it.content[0].type === "paragraph") {
      // Flatten single-paragraph items to match the server renderer.
      parts.push(renderInlines(it.content[0].content, scope, ctx));
    } else {
      for (const c of it.content) parts.push(renderBlock(c, scope, ctx));
    }
    parts.push("</li>\n");
  }
  parts.push(`</${tag}>\n`);
  return parts.join("");
}

function renderTable(
  rows: TableRow[],
  id: string | undefined,
  scope: Scope,
  ctx: RenderContext,
): string {
  if (rows.length === 0) return `<table${nodeAttr(id)}></table>\n`;
  const parts: string[] = [`<table${nodeAttr(id)}>\n`];
  const firstIsHeader = !!rows[0].header;
  if (firstIsHeader) {
    parts.push("<thead>");
    parts.push(renderTableRow(rows[0], scope, ctx, true));
    parts.push("</thead>\n<tbody>\n");
    for (let i = 1; i < rows.length; i++) {
      parts.push(renderTableRow(rows[i], scope, ctx, false));
    }
    parts.push("</tbody>\n");
  } else {
    parts.push("<tbody>\n");
    for (const row of rows) {
      parts.push(renderTableRow(row, scope, ctx, !!row.header));
    }
    parts.push("</tbody>\n");
  }
  parts.push("</table>\n");
  return parts.join("");
}

function renderTableRow(
  row: TableRow,
  scope: Scope,
  ctx: RenderContext,
  isHeader: boolean,
): string {
  const tag = isHeader ? "th" : "td";
  const parts: string[] = [`<tr${nodeAttr(row.id)}>`];
  for (const c of row.cells) {
    const cs = c.colSpan && c.colSpan > 1 ? ` colspan="${c.colSpan}"` : "";
    parts.push(`<${tag}${nodeAttr(c.id)}${cs}>`);
    parts.push(renderInlines(c.content, scope, ctx));
    parts.push(`</${tag}>`);
  }
  parts.push("</tr>\n");
  return parts.join("");
}

function renderImage(
  b: { src: string; alt?: string; width?: number; id?: string },
  ctx: RenderContext,
): string {
  if (!b.src) {
    pushDiag(ctx, "warning", "image block missing src", b.id);
    return "";
  }
  const w = b.width && b.width > 0 ? ` width="${b.width}"` : "";
  return `<img${nodeAttr(b.id)} src="${escapeAttr(b.src)}" alt="${escapeAttr(b.alt ?? "")}"${w} />\n`;
}

function renderRepeat(
  b: { source: string; as?: string; children: Block[]; id?: string },
  scope: Scope,
  ctx: RenderContext,
): string {
  if (!b.source) {
    pushDiag(ctx, "error", "repeat block missing source path", b.id);
    return "";
  }
  const raw = resolvePath(scope, b.source);
  const items = iterItems(raw);
  if (!items) return "";
  const parts: string[] = [];
  for (const item of items) {
    const inner = pushScope(scope, item, b.as ?? "");
    for (const c of b.children) parts.push(renderBlock(c, inner, ctx));
  }
  return parts.join("");
}

function renderIf(
  b: { condition: Condition; children: Block[]; else?: Block[]; id?: string },
  scope: Scope,
  ctx: RenderContext,
): string {
  if (!b.condition) {
    pushDiag(ctx, "error", "if block missing condition", b.id);
    return "";
  }
  const branch = evalCondition(b.condition, scope) ? b.children : (b.else ?? []);
  return branch.map((c) => renderBlock(c, scope, ctx)).join("");
}

// ---------------------------------------------------------------------------
// Inline rendering
// ---------------------------------------------------------------------------

function renderInlines(inl: Inline[], scope: Scope, ctx: RenderContext): string {
  const parts: string[] = [];
  for (const i of inl) {
    switch (i.type) {
      case "text":
        parts.push(writeTextWithMarks(i.text, i.marks));
        break;
      case "hardBreak":
        parts.push("<br />");
        break;
      case "field":
        parts.push(renderField(i, scope, ctx));
        break;
    }
  }
  return parts.join("");
}

function writeTextWithMarks(text: string, marks?: Mark[]): string {
  if (!text && (!marks || marks.length === 0)) return "";
  const { opens, closes } = marksToTags(marks ?? []);
  const body = escapeText(text);
  // Close in reverse so tags nest correctly.
  return opens.join("") + body + closes.slice().reverse().join("");
}

function marksToTags(marks: Mark[]): { opens: string[]; closes: string[] } {
  const opens: string[] = [];
  const closes: string[] = [];
  for (const m of marks) {
    switch (m.type) {
      case "bold":
        opens.push("<strong>");
        closes.push("</strong>");
        break;
      case "italic":
        opens.push("<em>");
        closes.push("</em>");
        break;
      case "underline":
        opens.push("<u>");
        closes.push("</u>");
        break;
      case "strike":
        opens.push("<s>");
        closes.push("</s>");
        break;
      case "code":
        opens.push("<code>");
        closes.push("</code>");
        break;
      case "link":
        opens.push(`<a href="${escapeAttr(m.href || "#")}">`);
        closes.push("</a>");
        break;
      case "color":
        if (m.value) {
          opens.push(`<span style="color:${escapeAttr(m.value)}">`);
          closes.push("</span>");
        }
        break;
    }
  }
  return { opens, closes };
}

function renderField(
  i: { path: string; format?: FieldFormat; default?: string },
  scope: Scope,
  ctx: RenderContext,
): string {
  const v = resolvePath(scope, i.path);
  if (v === undefined) {
    pushDiag(ctx, "warning", `field "${i.path}" not found in data`);
  }
  const out = formatField(v, i.format, i.default, ctx.locale);
  return `<span data-formly-field="${escapeAttr(i.path)}">${escapeText(out)}</span>`;
}

// ---------------------------------------------------------------------------
// Escaping helpers
// ---------------------------------------------------------------------------

function escapeText(s: string): string {
  // html.EscapeString-compatible: & < > ' "
  return s
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/'/g, "&#39;")
    .replace(/"/g, "&#34;");
}

function escapeAttr(s: string): string {
  return escapeText(s);
}

function nodeAttr(id: string | undefined): string {
  if (!id) return "";
  return ` data-formly-node="${escapeAttr(id)}"`;
}

// ---------------------------------------------------------------------------
// Preview helpers
// ---------------------------------------------------------------------------

/** Wrap a rendered body in the preview document shell.  Must stay visually
 *  aligned with the Go side's wrapInDocument; small deltas are OK (class
 *  names, commented regions) but typography and spacing should match.
 *
 *  `themeCss`, when provided, is appended AFTER the base stylesheet so
 *  per-template styles override shell defaults. Pass it through verbatim
 *  from `StoredDoc.themeCss`. */
export function wrapInDocument(body: string, themeCss?: string): string {
  const theme =
    themeCss && themeCss.trim() !== ""
      ? `<style data-formly-theme>\n${themeCss}\n</style>\n`
      : "";
  return `<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8" />
<style>
  :root { color-scheme: light; }
  body {
    font-family: Inter, system-ui, -apple-system, "Segoe UI", sans-serif;
    color: #111;
    max-width: 760px;
    margin: 0 auto;
    padding: 48px;
    line-height: 1.65;
    font-size: 14px;
  }
  h1, h2, h3, h4 { line-height: 1.25; margin-top: 1.4em; }
  h1 { font-size: 28px; border-bottom: 2px solid #e5e7eb; padding-bottom: 6px; }
  h2 { font-size: 22px; }
  h3 { font-size: 18px; }
  p { margin: 0.6em 0; }
  table { border-collapse: collapse; width: 100%; margin: 1em 0; }
  th, td { border: 1px solid #e5e7eb; padding: 8px 10px; text-align: left; }
  th { background: #f9fafb; }
  hr { border: 0; border-top: 1px solid #e5e7eb; margin: 2em 0; }
  a { color: #0d9488; }
  img { max-width: 100%; }
  ul, ol { padding-left: 1.5em; }
  [data-formly-field] { background: transparent; }
  @media print { body { padding: 24px; max-width: none; } }
</style>
${theme}</head>
<body>
${body}
</body>
</html>`;
}

File: api/internal/generate/doc/render.go
package doc

import (
	"fmt"
	"html"
	"strings"

	"github.com/docforge/api/internal/i18n"
)

// RenderContext is the ambient state threaded through the walker. It's
// mutable in one direction only: we append to `diagnostics` during the walk.
// The caller receives the final list; a renderer never panics or aborts on
// a single bad node — we emit a best-effort placeholder and record the
// problem so the editor can surface it.
type RenderContext struct {
	Locale      string
	I18n        i18n.Config
	Diagnostics []Diagnostic
}

// Diagnostic is a non-fatal problem encountered during render. The editor
// uses `NodeID` to outline the offending element in red.
type Diagnostic struct {
	NodeID   string `json:"nodeId,omitempty"`
	Severity string `json:"severity"` // "error" | "warning"
	Message  string `json:"message"`
}

// RenderDoc walks an AST and returns the rendered HTML (<body>-inner only —
// the caller wraps in a document shell via wrapInDocument).
//
// The renderer never returns an error. Problems are captured as diagnostics
// and embedded as comments in the output HTML so they're visible in dev-tools
// but don't break the visual layout. This matches the "best-effort render"
// contract already established by the existing html / markdown packages.
func RenderDoc(d *Doc, data map[string]interface{}, ctx *RenderContext) (string, []Diagnostic) {
	if ctx == nil {
		ctx = &RenderContext{}
	}
	if d == nil {
		return "", ctx.Diagnostics
	}
	var sb strings.Builder
	scp := newRootScope(data)
	for _, b := range d.Content {
		renderBlock(&sb, b, scp, ctx)
	}
	return sb.String(), ctx.Diagnostics
}

// renderBlock dispatches on the block discriminator. Unknown kinds emit an
// HTML comment rather than silently disappearing, so authors don't lose
// content to typos in older docs.
func renderBlock(sb *strings.Builder, b Block, s *scope, ctx *RenderContext) {
	switch b.Kind {
	case BlockHeading:
		renderHeading(sb, b, s, ctx)
	case BlockParagraph:
		renderParagraph(sb, b, s, ctx)
	case BlockBulletList:
		renderList(sb, b, s, ctx, "ul")
	case BlockOrderedList:
		renderList(sb, b, s, ctx, "ol")
	case BlockTable:
		renderTable(sb, b, s, ctx)
	case BlockDivider:
		sb.WriteString("<hr />\n")
	case BlockImage:
		renderImage(sb, b, ctx)
	case BlockRepeat:
		renderRepeat(sb, b, s, ctx)
	case BlockIf:
		renderIf(sb, b, s, ctx)
	case BlockRaw:
		// Intentionally unescaped — this is the documented escape hatch.
		sb.WriteString(b.HTML)
	default:
		ctx.pushDiag(b.ID, "warning",
			fmt.Sprintf("unknown block type %q", string(b.Kind)))
		fmt.Fprintf(sb, "<!-- unknown block type %q -->\n", string(b.Kind))
	}
}

func renderHeading(sb *strings.Builder, b Block, s *scope, ctx *RenderContext) {
	lvl := b.Level
	if lvl < 1 || lvl > 6 {
		lvl = 1
	}
	fmt.Fprintf(sb, "<h%d%s>", lvl, dataNodeAttr(b.ID))
	renderInlines(sb, b.Inline, s, ctx)
	fmt.Fprintf(sb, "</h%d>\n", lvl)
}

func renderParagraph(sb *strings.Builder, b Block, s *scope, ctx *RenderContext) {
	style := ""
	switch b.Align {
	case "center":
		style = ` style="text-align:center"`
	case "right":
		style = ` style="text-align:right"`
	case "left":
		style = ` style="text-align:left"`
	}
	fmt.Fprintf(sb, "<p%s%s>", dataNodeAttr(b.ID), style)
	renderInlines(sb, b.Inline, s, ctx)
	sb.WriteString("</p>\n")
}

func renderList(sb *strings.Builder, b Block, s *scope, ctx *RenderContext, tag string) {
	fmt.Fprintf(sb, "<%s%s>\n", tag, dataNodeAttr(b.ID))
	for _, it := range b.Items {
		fmt.Fprintf(sb, "<li%s>", dataNodeAttr(it.ID))
		// Inline the first paragraph-only child so simple lists don't get
		// a stray <p> wrapper that bloats the markup.
		if len(it.Content) == 1 && it.Content[0].Kind == BlockParagraph {
			renderInlines(sb, it.Content[0].Inline, s, ctx)
		} else {
			for _, c := range it.Content {
				renderBlock(sb, c, s, ctx)
			}
		}
		sb.WriteString("</li>\n")
	}
	fmt.Fprintf(sb, "</%s>\n", tag)
}

func renderTable(sb *strings.Builder, b Block, s *scope, ctx *RenderContext) {
	fmt.Fprintf(sb, "<table%s>\n", dataNodeAttr(b.ID))
	// Treat the first row as a header when explicitly flagged; otherwise
	// render a single-body table. (Inferring header-ness would risk
	// surprising authors who intentionally have data in row 0.)
	firstIsHeader := len(b.Rows) > 0 && b.Rows[0].Header
	if firstIsHeader {
		sb.WriteString("<thead>")
		renderTableRow(sb, b.Rows[0], s, ctx, true)
		sb.WriteString("</thead>\n<tbody>\n")
		for _, row := range b.Rows[1:] {
			renderTableRow(sb, row, s, ctx, false)
		}
		sb.WriteString("</tbody>\n")
	} else {
		sb.WriteString("<tbody>\n")
		for _, row := range b.Rows {
			renderTableRow(sb, row, s, ctx, row.Header)
		}
		sb.WriteString("</tbody>\n")
	}
	sb.WriteString("</table>\n")
}

func renderTableRow(sb *strings.Builder, row TableRow, s *scope, ctx *RenderContext, isHeader bool) {
	fmt.Fprintf(sb, "<tr%s>", dataNodeAttr(row.ID))
	tag := "td"
	if isHeader {
		tag = "th"
	}
	for _, c := range row.Cells {
		fmt.Fprintf(sb, "<%s%s", tag, dataNodeAttr(c.ID))
		if c.ColSpan > 1 {
			fmt.Fprintf(sb, ` colspan="%d"`, c.ColSpan)
		}
		sb.WriteByte('>')
		renderInlines(sb, c.Inline, s, ctx)
		fmt.Fprintf(sb, "</%s>", tag)
	}
	sb.WriteString("</tr>\n")
}

func renderImage(sb *strings.Builder, b Block, ctx *RenderContext) {
	if b.Src == "" {
		ctx.pushDiag(b.ID, "warning", "image block missing src")
		return
	}
	widthAttr := ""
	if b.Width > 0 {
		widthAttr = fmt.Sprintf(` width="%d"`, b.Width)
	}
	fmt.Fprintf(sb, "<img%s src=\"%s\" alt=\"%s\"%s />\n",
		dataNodeAttr(b.ID),
		html.EscapeString(b.Src),
		html.EscapeString(b.Alt),
		widthAttr,
	)
}

// renderRepeat iterates the source path and emits the inner blocks per item.
// A missing or non-iterable source renders nothing — the same semantics as
// Go templates' `range` over missing data.
func renderRepeat(sb *strings.Builder, b Block, s *scope, ctx *RenderContext) {
	if b.Source == "" {
		ctx.pushDiag(b.ID, "error", "repeat block missing source path")
		return
	}
	raw, _ := s.resolve(b.Source)
	items := iterItems(raw)
	if items == nil {
		return
	}
	for _, item := range items {
		inner := s.push(item, b.As)
		for _, c := range b.Children {
			renderBlock(sb, c, inner, ctx)
		}
	}
}

// renderIf evaluates the condition and emits either the `children` branch or
// the `else` branch. An absent-else with a false condition emits nothing.
func renderIf(sb *strings.Builder, b Block, s *scope, ctx *RenderContext) {
	if b.Condition == nil {
		ctx.pushDiag(b.ID, "error", "if block missing condition")
		return
	}
	branch := b.Children
	if !evalCondition(b.Condition, s) {
		branch = b.Else
	}
	for _, c := range branch {
		renderBlock(sb, c, s, ctx)
	}
}

// renderInlines emits an inline-node run. Text nodes are HTML-escaped and
// wrapped in any applicable mark tags (bold/italic/link/etc); fields are
// resolved against the current scope and formatted.
func renderInlines(sb *strings.Builder, inl []Inline, s *scope, ctx *RenderContext) {
	for _, i := range inl {
		switch i.Kind {
		case InlineText, "":
			writeTextWithMarks(sb, i.Text, i.Marks)
		case InlineHardBreak:
			sb.WriteString("<br />")
		case InlineField:
			renderField(sb, i, s, ctx)
		default:
			ctx.pushDiag("", "warning",
				fmt.Sprintf("unknown inline type %q", string(i.Kind)))
		}
	}
}

// writeTextWithMarks wraps the escaped text in open/close tags for each mark.
// Marks are applied in a stable order so the output is deterministic — the
// order doesn't matter semantically for most combinations, but tests + diff
// tooling like reproducibility.
func writeTextWithMarks(sb *strings.Builder, text string, marks []Mark) {
	if text == "" && len(marks) == 0 {
		return
	}
	opens, closes := marksToTags(marks)
	for _, o := range opens {
		sb.WriteString(o)
	}
	sb.WriteString(html.EscapeString(text))
	// Close in reverse so tags nest correctly.
	for i := len(closes) - 1; i >= 0; i-- {
		sb.WriteString(closes[i])
	}
}

// marksToTags translates the mark list into paired open/close strings.
// Returning two slices keeps the caller symmetric and avoids a stack
// allocation per inline run.
func marksToTags(marks []Mark) (opens, closes []string) {
	// Stable iteration order: the AST-defined order of appearance.
	for _, m := range marks {
		switch m.Type {
		case MarkBold:
			opens = append(opens, "<strong>")
			closes = append(closes, "</strong>")
		case MarkItalic:
			opens = append(opens, "<em>")
			closes = append(closes, "</em>")
		case MarkUnderline:
			opens = append(opens, "<u>")
			closes = append(closes, "</u>")
		case MarkStrike:
			opens = append(opens, "<s>")
			closes = append(closes, "</s>")
		case MarkCode:
			opens = append(opens, "<code>")
			closes = append(closes, "</code>")
		case MarkLink:
			href := m.Href
			if href == "" {
				href = "#"
			}
			opens = append(opens,
				fmt.Sprintf(`<a href="%s">`, html.EscapeString(href)))
			closes = append(closes, "</a>")
		case MarkColor:
			if m.Value != "" {
				opens = append(opens,
					fmt.Sprintf(`<span style="color:%s">`, html.EscapeString(m.Value)))
				closes = append(closes, "</span>")
			}
		}
	}
	return opens, closes
}

// renderField resolves a field node and writes the escaped, formatted value.
// Missing values emit the declared `default`, or an empty string if none.
// The rendered span carries data attributes so the editor can round-trip
// the field definition when re-parsing saved-to-HTML previews.
func renderField(sb *strings.Builder, i Inline, s *scope, ctx *RenderContext) {
	v, ok := s.resolve(i.Path)
	if !ok {
		// Surface a diagnostic but don't fail — authors may bind this later.
		ctx.pushDiag("", "warning", fmt.Sprintf("field %q not found in data", i.Path))
	}
	formatted := formatField(v, i.Format, i.Default, ctx.Locale, ctx.I18n)
	// Wrap in a span tagged with the path so the preview stays visually
	// attributable (and so CSS in the preview can highlight bound fields).
	fmt.Fprintf(sb, `<span data-formly-field="%s">`, html.EscapeString(i.Path))
	sb.WriteString(html.EscapeString(formatted))
	sb.WriteString("</span>")
}

// dataNodeAttr returns ` data-formly-node="ID"` or "" when the node has no
// ID. Centralizing this keeps the attribute name in one place and avoids
// emitting ` data-formly-node=""` noise for ID-less nodes.
func dataNodeAttr(id string) string {
	if id == "" {
		return ""
	}
	return fmt.Sprintf(` data-formly-node="%s"`, html.EscapeString(id))
}

// pushDiag appends a diagnostic. Lives on *RenderContext rather than as a
// free function because future debugging may want to mutate more than just
// Diagnostics (e.g., a counter of skipped nodes).
func (c *RenderContext) pushDiag(nodeID, severity, msg string) {
	c.Diagnostics = append(c.Diagnostics, Diagnostic{
		NodeID: nodeID, Severity: severity, Message: msg,
	})
}
